pkg/gadgets/biolatency/tracer/core/tracer.go

```diff
@@ -0,0 +1,140 @@
+//go:build linux
+// +build linux
+
+// Copyright 2019-2021 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tracer
+
+// #include <linux/types.h>
+// #include "./bpf/biolatency.h"
+import "C"
+
+import (
+	"encoding/json"
+	"fmt"
+	"unsafe"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/link"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets/biolatency/types"
+)
+
+//go:generate sh -c "GOOS=$(go env GOHOSTOS) GOARCH=$(go env GOHOSTARCH) go run github.com/cilium/ebpf/cmd/bpf2go -target bpfel -cc clang biolatency ./bpf/biolatency.bpf.c -- -I./bpf/ -I../../../../ -target bpf -D__TARGET_ARCH_x86"
+
+type Tracer struct {
+	node                string
+	objs                biolatencyObjects
+	blockRqCompleteLink link.Link
+	blockRqInsertLink   link.Link
+	blockRqIssueLink    link.Link
+}
+
+func NewTracer(node string) (*Tracer, error) {
+	t := &Tracer{
+		node: node,
+	}
+
+	if err := t.start(); err != nil {
+		t.Stop()
+		return nil, err
+	}
+
+	return t, nil
+}
+
+func getReport(histMap *ebpf.Map) (types.Report, error) {
+	report := types.Report{
+		ValType: "usecs",
+		Time:    "foo",
+	}
+
+	key := C.struct_hist_key{}
+
+	err := histMap.NextKey(nil, unsafe.Pointer(&key))
+	if err != nil {
+		return types.Report{}, fmt.Errorf("error getting next key: %w", err)
+	}
+
+	hist := C.struct_hist{}
+	if err := histMap.Lookup(key, unsafe.Pointer(&hist)); err != nil {
+		return types.Report{}, err
+	}
+
+	data := []types.Data{}
+	indexMax := 0
+	for i, val := range hist.slots {
+		if val > 0 {
+			indexMax = i
+		}
+
+		data = append(data, types.Data{
+			Count:         uint64(val),
+			IntervalStart: (uint64(1) << (i + 1)) >> 1,
+			IntervalEnd:   (uint64(1) << (i + 1)) - 1,
+		})
+	}
+
+	report.Data = data[:indexMax]
+
+	return report, nil
+}
+
+func (t *Tracer) Stop() (string, error) {
+	t.blockRqCompleteLink = gadgets.CloseLink(t.blockRqCompleteLink)
+	t.blockRqInsertLink = gadgets.CloseLink(t.blockRqInsertLink)
+	t.blockRqIssueLink = gadgets.CloseLink(t.blockRqIssueLink)
+
+	report, err := getReport(t.objs.biolatencyMaps.Hists)
```

Comment:
This can cause crashes if `Stop()` is called before this map is created. (It could happen if the program fails to load on L112 below). An easy solution is to add a conditional checking if `t.objs.Hists != nil`, it's not cleanest way but I think it's fine.